WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -3,60 +3,188 @@
+				if (outputCount == ownOutputCount)
+				{
+					AnalyzeSelfSpend(tx);
+				}
+				else
+				{
+					AnalyzeNormalSpend(tx);
+				}
+			}
+			else
+			{
+				AnalyzeCoinjoin(tx);
+			}
+
+			AnalyzeClusters(tx);
+		}
+
+		private void AnalyzeCoinjoin(SmartTransaction tx)
+		{
+			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset);
+
+			var indistinguishableWalletOutputs = tx
+				.WalletOutputs.GroupBy(x => x.Amount)
+				.ToDictionary(x => x.Key, y => y.Count());
+
+			foreach (var newCoin in tx.WalletOutputs.ToArray())
 			{
-				// Calculate anonymity sets.
 				// Get the anonymity set of i-th output in the transaction.
 				var anonset = tx.Transaction.GetAnonymitySet(newCoin.Index);
 
-				// If we provided inputs to the transaction.
-				if (tx.WalletInputs.Any())
+				// Don't create many anonset if we've provided a lot of equal outputs.
+				anonset -= indistinguishableWalletOutputs[newCoin.Amount] - 1;
+
+				// We should only consider ourselves once in the anonset.
+				anonset += newInputAnonset - 1;
+
+				HdPubKey hdPubKey = newCoin.HdPubKey;
+				if (hdPubKey.AnonymitySet == HdPubKey.DefaultHighAnonymitySet)
+				{
+					// If the new coin's HD pubkey haven't been used yet
+					// then its anonset haven't been set yet.
+					// In that case the acquired anonset does not have to be intersected with the default anonset,
+					// so this coin gets the aquired anonset.
+					hdPubKey.AnonymitySet = anonset;
+				}
+				else if (distinctWalletInputPubKeys.Contains(hdPubKey))
 				{
-					// Take the input that we provided with the smallest anonset.
-					// And add that to the base anonset from the tx.
-					// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
-					// Take minus 1, because we do not want to count own into the anonset, so...
-					// If the anonset of our UTXO would be 1, and the smallest anonset of our inputs would be 1, too, then we don't make...
-					// The new UTXO's anonset 2, but only 1.
-					anonset += tx.WalletInputs.Select(x => x.HdPubKey).Min(x => x.AnonymitySet) - 1;
+					// If it's a reuse of an input's pubkey, then intersection punishment is senseless.
+					hdPubKey.AnonymitySet = newInputAnonset;
 				}
+				else if (tx.WalletOutputs.Where(x => x != newCoin).Select(x => x.HdPubKey).Contains(hdPubKey))
+				{
+					// If it's a reuse of another output' pubkey, then intersection punishment can only go as low as the inherited anonset.
+					hdPubKey.AnonymitySet = Math.Max(newInputAnonset, Intersect(new[] { anonset, hdPubKey.AnonymitySet }, 1));
+				}
+				else
+				{
+					hdPubKey.AnonymitySet = Intersect(new[] { anonset, hdPubKey.AnonymitySet }, 1);
+				}
+			}
+		}
+
+		/// <param name="newInputAnonset">The new anonymity set of the inputs.</param>
+		private void AnalyzeWalletInputs(SmartTransaction tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset)
```

Comment:
It was previously `inheritedAnonSet`. But that lead to a bunch of +-1 checks, see @molnard's comment here: https://github.com/zkSNACKs/WalletWasabi/pull/4724#discussion_r524073785

Then I realized if it doesn't return anonset intended to be inherited by outputs, but rather it would return more generally the new anonset of the inputs (which is unified at this point due to intersection punishment)  then we can remove a bunch of explanation comments with the +-1 checks.